frontend/src/atoms/app.ts

```diff
@@ -0,0 +1,57 @@
+import { urlSearchParamsStorage } from '@/lib/storage';
+import { UIResponseProto } from '@/types/generated/ui';
+import { StartupResponse } from '@/types/startup';
+import type { FeatureCollection } from 'geojson';
+import { atomWithImmer } from 'jotai-immer';
+import { atomWithStorage } from 'jotai/utils';
+import { Point } from 'maplibre-gl';
+
+type Scenario = {
+    name: string;
+};
+
+type Scenarios = Record<string, Scenario>;
+
+export const collectionAtom = atomWithStorage(
+    'r',
+    '',
+    urlSearchParamsStorage({}),
+    {
+        getOnInit: true,
+    }
+);
+
+export type AppStore = {
+    startup?: StartupResponse;
+    scenarios: Scenarios;
+    tabs: {
+        left: keyof Scenarios;
+        right?: keyof Scenarios;
+    };
+    geojson: Record<string, FeatureCollection[]>;
+    stacks: Record<
+        string,
+        {
+            id: string;
+            expression?: string;
+            coordinates?: Point;
+            docked: boolean;
+            transient?: boolean;
+            proto: UIResponseProto;
+            tab?: keyof Scenarios;
+        }
+    >;
+};
+
+export const appAtom = atomWithImmer<AppStore>({
```

Comment:
This object aggregates a lot of different things, which isn't good for encapsulation (ie the code handling one stack could accidentally modify the state of another). Is there an argument for breaking it up? Also - this is probably not the case, as I'm new to this react world, but does this not mean that a change to any stack triggers a re-render of all stacks, since everything depends on the same atom?

Reply:
To be honest, it's something I'm still considering. I like keeping some global state in applications with plenty of interaction that impacts different parts of the application. With it, changes to the application state can be centralized and it avoids a lot of prop drilling. However, it can indeed make this more complicated for debugging. So, I'm not 100% set on what should be global and what should be local, at this moment. With the global state, the goal is to create more granular getters and reducers to minimize direct interactions with it.

As for the second question, you're right that changes to a plain object would trigger an update. But with Immer, any unchanged parts of the state retain the same reference and avoid unnecessary updates [https://immerjs.github.io/immer](https://immerjs.github.io/immer/). That's how we keep our global state efficient. Unfortunately, there are still some re-renders happening because I'm often getting the full app object. But, I'm working on creating those more specific functions I mentioned before that should improve this.